File: georocket-server/src/main/java/io/georocket/output/MultiMerger.java
package io.georocket.output;

import io.georocket.output.geojson.GeoJsonMerger;
import io.georocket.output.xml.XMLMerger;
import io.georocket.storage.ChunkMeta;
import io.georocket.storage.ChunkReadStream;
import io.georocket.storage.GeoJsonChunkMeta;
import io.georocket.storage.XMLChunkMeta;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.streams.WriteStream;
import rx.Completable;

/**
 * <p>A merger that either delegates to {@link XMLMerger} or
 * {@link GeoJsonMerger} depending on the types of the chunks to merge.</p>
 * <p>For the time being the merger can only merge chunks of the same type.
 * In the future it may create an archive (e.g. a ZIP or a TAR file) containing
 * chunks of mixed types.</p>
 * @author Michel Kraemer
 */
public class MultiMerger implements Merger<ChunkMeta> {
  private XMLMerger xmlMerger;
  private GeoJsonMerger geoJsonMerger;

  /**
   * {@code true} if chunks should be merged optimistically without
   * prior initialization
   */
  private final boolean optimistic;

  /**
   * Creates a new merger
   * @param optimistic {@code true} if chunks should be merged optimistically
   * without prior initialization
   */
  public MultiMerger(boolean optimistic) {
    this.optimistic = optimistic;
  }
  
  private Completable ensureMerger(ChunkMeta meta) {
    if (meta instanceof XMLChunkMeta) {
      if (xmlMerger == null) {
        if (geoJsonMerger != null) {
          return Completable.error(new IllegalStateException("Cannot merge "
            + "XML chunk into a GeoJSON document."));
        }
        xmlMerger = new XMLMerger(optimistic);
      }
      return Completable.complete();
    } else if (meta instanceof GeoJsonChunkMeta) {
      if (geoJsonMerger == null) {
        if (xmlMerger != null) {
          return Completable.error(new IllegalStateException("Cannot merge "
            + "GeoJSON chunk into an XML document."));
        }
        geoJsonMerger = new GeoJsonMerger(optimistic);
      }
      return Completable.complete();
    }
    return Completable.error(new IllegalStateException("Cannot merge "
      + "chunk of type " + meta.getMimeType()));
  }
  
  @Override
  public Completable init(ChunkMeta meta) {
    return ensureMerger(meta)
      .andThen(Completable.defer(() -> {
        if (meta instanceof XMLChunkMeta) {
          return xmlMerger.init((XMLChunkMeta)meta);
        }
        return geoJsonMerger.init((GeoJsonChunkMeta)meta);
      }));
  }

  @Override
  public Completable merge(ChunkReadStream chunk, ChunkMeta meta,
      WriteStream<Buffer> out) {
    return ensureMerger(meta)
      .andThen(Completable.defer(() -> {
        if (meta instanceof XMLChunkMeta) {
          return xmlMerger.merge(chunk, (XMLChunkMeta)meta, out);
        }
        return geoJsonMerger.merge(chunk, (GeoJsonChunkMeta)meta, out);
      }));
  }

  @Override
  public void finish(WriteStream<Buffer> out) {
    if (xmlMerger != null) {
      xmlMerger.finish(out);
    }
    if (geoJsonMerger != null) {
      geoJsonMerger.finish(out);
    }
  }
}

File: georocket-server/src/main/kotlin/io/georocket/output/MultiMerger.kt
package io.georocket.output

import io.georocket.storage.ChunkMeta
import io.georocket.output.xml.XMLMerger
import io.georocket.output.geojson.GeoJsonMerger
import io.georocket.storage.XMLChunkMeta
import java.lang.IllegalStateException
import io.georocket.storage.GeoJsonChunkMeta
import io.vertx.core.buffer.Buffer
import io.vertx.core.streams.WriteStream

/**
 *
 * A merger that either delegates to [XMLMerger] or
 * [GeoJsonMerger] depending on the types of the chunks to merge.
 *
 * For the time being the merger can only merge chunks of the same type.
 * In the future, it may create an archive (e.g. a ZIP or a TAR file) containing
 * chunks of mixed types.
 * @param optimistic `true` if chunks should be merged optimistically
 * without prior initialization
 * @author Michel Kraemer
 */
class MultiMerger(private val optimistic: Boolean) : Merger<ChunkMeta> {
  private var xmlMerger: XMLMerger? = null
  private var geoJsonMerger: GeoJsonMerger? = null

  private fun ensureMerger(chunkMetadata: ChunkMeta) {
    when (chunkMetadata) {
      is XMLChunkMeta -> if (xmlMerger == null) {
        if (geoJsonMerger != null) {
          throw IllegalStateException("Cannot merge XML chunk into a GeoJSON document.")
        }
        xmlMerger = XMLMerger(optimistic)
      }

      is GeoJsonChunkMeta -> if (geoJsonMerger == null) {
        if (xmlMerger != null) {
          throw IllegalStateException("Cannot merge GeoJSON chunk into an XML document.")
        }
        geoJsonMerger = GeoJsonMerger(optimistic)
      }

      else -> throw IllegalStateException("Cannot merge chunk of type ${chunkMetadata.mimeType}")
    }
  }

  override fun init(chunkMetadata: ChunkMeta) {
    ensureMerger(chunkMetadata)

    if (chunkMetadata is XMLChunkMeta) {
      xmlMerger!!.init(chunkMetadata)
    } else {
      geoJsonMerger!!.init(chunkMetadata as GeoJsonChunkMeta)
    }
  }

  override suspend fun merge(chunk: Buffer, chunkMetadata: ChunkMeta,
      outputStream: WriteStream<Buffer>) {
    ensureMerger(chunkMetadata)

    if (chunkMetadata is XMLChunkMeta) {
      xmlMerger!!.merge(chunk, chunkMetadata, outputStream)
    } else {
      geoJsonMerger!!.merge(chunk, chunkMetadata as GeoJsonChunkMeta, outputStream)
    }
  }

  override fun finish(outputStream: WriteStream<Buffer>) {
    if (xmlMerger != null) {
      xmlMerger!!.finish(outputStream)
    }
    if (geoJsonMerger != null) {
      geoJsonMerger!!.finish(outputStream)
    }
  }
}
